### PR TITLE
Create LRA blog

### PR DESCRIPTION
**Pull request by yongja79**
Link to draft site https://draft-openlibertyio.mybluemix.net/blog/2020/12/18/microprofile-long-running-actions-beta.html
Blog issue: #926

### CODE REVIEW

File: posts/2020-12-18-microprofile-long-running-actions-beta.adoc [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+title: "Microprofile Long Running Actions in Open Liberty"
+categories: blog
+author_picture: https://avatars3.githubusercontent.com/yongja79
+author_github: https://github.com/yongja79
+seo-title: Microprofile Long Running Actions in Open Liberty
+seo-description: Try Microprofile Long Running Actions in the Open Liberty 20.0.0.12-beta
+blog_description: "Try Microprofile Long Running Actions in the Open Liberty 20.0.0.12-beta"
+open-graph-image: https://openliberty.io/img/twitter_card.jpg
+---
+= Microprofile Long Running Actions in Open Liberty
```
Comment (Emily-Jiang) [inline posts/2020-12-18-microprofile-long-running-actions-beta.adoc]: Please replace all Microprofile with MicroProfile

File: posts/2020-12-18-microprofile-long-running-actions-beta.adoc [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+Jason Yong <https://github.com/yongja79>
+
+In the 20.0.0.12-beta release of Open Liberty a new capability was introduced to support loosely-coupled transaction semantics for Java microservice. This is an implementation of a new MicroProfile specification for Long Running Actions (LRAs).
```
Comment (Emily-Jiang) [inline posts/2020-12-18-microprofile-long-running-actions-beta.adoc]: (LRAs) -> (MicroProfile LRA)

File: posts/2020-12-18-microprofile-long-running-actions-beta.adoc [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+seo-description: Try Microprofile Long Running Actions in the Open Liberty 20.0.0.12-beta
+blog_description: "Try Microprofile Long Running Actions in the Open Liberty 20.0.0.12-beta"
+open-graph-image: https://openliberty.io/img/twitter_card.jpg
+---
+= Microprofile Long Running Actions in Open Liberty
+Jason Yong <https://github.com/yongja79>
+
+In the 20.0.0.12-beta release of Open Liberty a new capability was introduced to support loosely-coupled transaction semantics for Java microservice. This is an implementation of a new MicroProfile specification for Long Running Actions (LRAs).  
+
+== What is a LRA?
```
Comment (Emily-Jiang) [inline posts/2020-12-18-microprofile-long-running-actions-beta.adoc]: What is a LRA? -> What is MicroProfile LRA?

File: posts/2020-12-18-microprofile-long-running-actions-beta.adoc [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+
+Long Running Actions is a MicroProfile specification that provides a simple, loosely-coupled transaction model for microservices and is based on the Saga pattern for distributed transactions . It relaxes some of the constraints of ACID transactions to enable independent microservices to more easily participate in a potentially long-running orchestrated activity, with each microservice contributing the encapsulated business logic required to align with the overall outcome of the activity or "transaction".
```
Comment (Emily-Jiang) [inline posts/2020-12-18-microprofile-long-running-actions-beta.adoc]: Long Running Actions -> MicroProfile LAR is

File: posts/2020-12-18-microprofile-long-running-actions-beta.adoc [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+* allow actions to finish early
+
+The LRA model relies on having compensating actions for all business interactions and ensures that when the activity ends either all the work is accepted or all will be compensated. How each activity gets compensated is down to the individual services, but the model defines what triggers a compensating action and when they are executed.
+
+== LRA in Open Liberty
+
+In Open Liberty there are 2 parts to the using LRA:
+
+* LRA Coordinator which manages the LRA processing. It handles the initialisation of the LRAs, enlisting in LRAs and the completion or compensating of LRAs. This is enabled through standard Liberty configuration.
+* LRA Participants which are applications using MicroProfile LRA annotations that cause them to become involved in the LRA
```
Comment (Emily-Jiang) [inline posts/2020-12-18-microprofile-long-running-actions-beta.adoc:40]: pleasae swap the order of these two items

File: posts/2020-12-18-microprofile-long-running-actions-beta.adoc [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+To try LRA in Open Liberty you need to download and extract the latest link:https://openliberty.io/downloads/#runtime_betas[beta driver]
+
+Create a new Open Liberty server to act as the coordinator by running the following command:
+
+[source, bash]
+----
+bin/server create LRACoordinator
+----
+
+In order to start a coordinator in Open Liberty you first need to enable the mpLRACoordinator-1.0 feature via the server.xml of the server as well as the cdi-2.0 and jaxrs-2.1 upon which it is dependant.
```
Comment (Emily-Jiang) [inline posts/2020-12-18-microprofile-long-running-actions-beta.adoc:54]: You can also use maven dependency as the beta is in maven central

Comment (Emily-Jiang) [reply posts/2020-12-18-microprofile-long-running-actions-beta.adoc:54]:
```
                    <runtimeArtifact>
                        <groupId>io.openliberty.beta</groupId>
                        <artifactId>openliberty-runtime</artifactId>
                        <version>20.0.0.12-beta</version>
                        <type>zip</type>
                    </runtimeArtifact>
```

File: posts/2020-12-18-microprofile-long-running-actions-beta.adoc [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+
+* LRA Coordinator which manages the LRA processing. It handles the initialisation of the LRAs, enlisting in LRAs and the completion or compensating of LRAs. This is enabled through standard Liberty configuration.
+* LRA Participants which are applications using MicroProfile LRA annotations that cause them to become involved in the LRA
+
+The typical setup would be to have a single coordinator on its own Open Liberty server, coordinating multiple participants which may be in a single Open Liberty server or across different servers.
+
+image::/img/blog/lra_typical_setup.png[LRA coordinator and participant setup,width=50%,align="center"]
+
+=== Starting a Microprofile LRA Coordinator in Open Liberty
+To try LRA in Open Liberty you need to download and extract the latest link:https://openliberty.io/downloads/#runtime_betas[beta driver]
```
Comment (Emily-Jiang) [inline posts/2020-12-18-microprofile-long-running-actions-beta.adoc]: You can also use maven dependency as the beta is in maven central
```
<runtimeArtifact>
                        <groupId>io.openliberty.beta</groupId>
                        <artifactId>openliberty-runtime</artifactId>
                        <version>20.0.0.12-beta</version>
                        <type>zip</type>
                    </runtimeArtifact>
```

File: posts/2020-12-18-microprofile-long-running-actions-beta.adoc [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+[AUDIT   ] CWWKZ0001I: Application mpLRACoordinator started in 8.045 seconds.
+----
+It is now ready to coordinate LRAs.
+
+=== Creating a participant service
+
+An LRA is started by the Liberty LRA coordinator when a participant service is annotated to require one. The coordinator creates a unique id for the LRA and makes it available to every participant in the LRA, enabling a participant to later register a compensating action for that specific LRA. All participant interactions with the LRA are done via annotations on participant methods.                                                                                                                                                                                       
+The most basic of LRAs has a single participant which needs 3 annotated methods:
+
+. Join/Create LRA method and do any business logic required.
```
Comment (Emily-Jiang) [inline posts/2020-12-18-microprofile-long-running-actions-beta.adoc]: Join/Create LRA method using `@LRA`

File: posts/2020-12-18-microprofile-long-running-actions-beta.adoc [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+. A complete method to be called once the LRA has completed successfully and do any business logic required.
```
Comment (Emily-Jiang) [inline posts/2020-12-18-microprofile-long-running-actions-beta.adoc]: A complete method using `@Complete`

File: posts/2020-12-18-microprofile-long-running-actions-beta.adoc [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+. A compensate method to be called should the LRA fail for any reason and include any logic required to revert any changes the join/create method did.
```
Comment (Emily-Jiang) [inline posts/2020-12-18-microprofile-long-running-actions-beta.adoc]: A compensate method using `@Compensate`

File: posts/2020-12-18-microprofile-long-running-actions-beta.adoc [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+        }
+        else {
+            System.out.println("Hotel booking failed");
+            return Response.serverError().build();
+        }
+    }
+----
+As an example of another LRA.Type the BookHotel has been set to MANDATORY which mean that it has to be called as part of an LRA and should it be called outside of an LRA it will automatically fail. So while the BookFlight service could be called outside of an LRA, as it would start its own, the BookHotel would return an error if that was attempted.
+
+It is usually best practice for each service to be deployed on a separate Open Liberty servers, but for convenience, deploy the BookHoliday.war and BookHotel.war to LRAParticipant and add the following lines to the server.xml
```
Comment (Emily-Jiang) [inline posts/2020-12-18-microprofile-long-running-actions-beta.adoc]: servers -> server

File: posts/2020-12-18-microprofile-long-running-actions-beta.adoc (hunk 3 above)
Comment (Emily-Jiang) [inline posts/2020-12-18-microprofile-long-running-actions-beta.adoc]: Can you add a section on why you need MP LRA? Why is 2-phase commit not suitable?

**Review by Emily-Jiang — CHANGES_REQUESTED** [review]

File: posts/2020-12-18-microprofile-long-running-actions-beta.adoc (hunk 3 above)
Comment (yongja79) [reply posts/2020-12-18-microprofile-long-running-actions-beta.adoc]: The why is actually in the "So how does MicroProfile LRA help?" section which had a comparison with 2PC in an earlier draft but was taken out by Ian R as he didn't feel it needed it.

File: posts/2020-12-18-microprofile-long-running-actions-beta.adoc [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+= MicroProfile Long Running Actions in Open Liberty
+Jason Yong <https://github.com/yongja79>
+
+In the 20.0.0.12-beta release of Open Liberty a new capability was introduced to support loosely-coupled transaction semantics for Java microservice. This is an implementation of a new MicroProfile specification for Long Running Actions (MicroProfile LRAs).  
+
+== What is a MicroProfile LRA?
+
+MicroProfile LRA is a MicroProfile specification that provides a simple, loosely-coupled transaction model for microservices and is based on the Saga pattern for distributed transactions . It relaxes some of the constraints of ACID transactions to enable independent microservices to more easily participate in a potentially long-running orchestrated activity, with each microservice contributing the encapsulated business logic required to align with the overall outcome of the activity or "transaction". 
+
+== So how does MicroProfile LRA help?
```
Comment (Emily-Jiang) [inline posts/2020-12-18-microprofile-long-running-actions-beta.adoc]: == Why do we need MicroProfile LRA?
In traditional applications, the well-known transaction model is ACID, where the most commonly used atomic commit protocol is  2-phase commit. In an ACID transaction, tightly-coupled XA resources can be directed to "rollback" by a coordinating transaction manager. This introduces resource locking and global rollback, which does not scale well and therefore not suitable for microservice architecture. As a consequence, [SAGA pattern](https://developer.ibm.com/depmodels/microservices/articles/use-saga-to-solve-distributed-transaction-management-problems-in-a-microservices-architecture#saga) was defined for microservice transaction architecture. A SAGA is a sequence of transactions. If one transaction fails, the saga undoes all of the preceding transactions. It aims for eventual consistence. MicroProfile LRA is an implementation of the SAGA pattern.

File: posts/2020-12-18-microprofile-long-running-actions-beta.adoc [hunk 13]
```diff
@@ -0,0 +1,400 @@
...
+
+== What is a MicroProfile LRA?
+
+MicroProfile LRA is a MicroProfile specification that provides a simple, loosely-coupled transaction model for microservices and is based on the Saga pattern for distributed transactions . It relaxes some of the constraints of ACID transactions to enable independent microservices to more easily participate in a potentially long-running orchestrated activity, with each microservice contributing the encapsulated business logic required to align with the overall outcome of the activity or "transaction". 
+
+== So how does MicroProfile LRA help? 
+
+A distributed microservices system can benefit from a form of transaction management to reduce the complexity of application-level error handling required when a multi-service business activity does not complete successfully. In an ACID transaction, tightly-coupled XA resources can be directed to "rollback" by a coordinating transaction manager. In a more loosely-coupled system of independent microservices then a compensating model can be more appropriate and this is what LRA provides.
```
Comment (Emily-Jiang) [inline posts/2020-12-18-microprofile-long-running-actions-beta.adoc]: Delete this paragraph

**Pull request closed without merging**